File: src/main/java/com/demonwav/mcdev/platform/mcp/gradle/McpDataService.java
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2016 Kyle Wood (DemonWav)
 *
 * MIT License
 */
package com.demonwav.mcdev.platform.mcp.gradle;

import com.demonwav.mcdev.buildsystem.gradle.AbstractDataService;
import com.demonwav.mcdev.platform.mcp.McpModuleType;

import com.intellij.openapi.externalSystem.model.DataNode;
import com.intellij.openapi.externalSystem.model.project.LibraryDependencyData;
import com.intellij.openapi.externalSystem.model.project.ProjectData;
import com.intellij.openapi.externalSystem.service.project.IdeModifiableModelsProvider;
import com.intellij.openapi.project.Project;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.plugins.gradle.util.GradleConstants;

import java.util.Collection;

public class McpDataService extends AbstractDataService {

    public McpDataService() {
        super(McpModuleType.getInstance());
    }

    @Override
    public void importData(@NotNull Collection<DataNode<LibraryDependencyData>> toImport,
                           @Nullable ProjectData projectData,
                           @NotNull Project project,
                           @NotNull IdeModifiableModelsProvider modelsProvider) {
        if (projectData == null || !projectData.getOwner().equals(GradleConstants.SYSTEM_ID)) {
            return;
        }

        checkModule(toImport, modelsProvider,McpModuleType.getInstance(), "minecraftSrc", "forgeSrc", "forgeBin", "minecraft_serverSrc", "minecraft_clientSrc");
    }
}

File: src/main/kotlin/com/demonwav/mcdev/platform/mcp/gradle/McpDataService.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2019 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.platform.mcp.gradle

import com.demonwav.mcdev.facet.MinecraftFacet
import com.demonwav.mcdev.platform.mcp.McpModuleType
import com.intellij.openapi.externalSystem.model.DataNode
import com.intellij.openapi.externalSystem.model.Key
import com.intellij.openapi.externalSystem.model.project.ProjectData
import com.intellij.openapi.externalSystem.service.project.IdeModifiableModelsProvider
import com.intellij.openapi.externalSystem.service.project.manage.AbstractProjectDataService
import com.intellij.openapi.module.Module
import com.intellij.openapi.project.Project

class McpDataService : AbstractProjectDataService<McpModelData, Module>() {

    override fun getTargetDataKey(): Key<McpModelData> = McpModelData.KEY

    override fun importData(
        toImport: Collection<DataNode<McpModelData>>,
        projectData: ProjectData?,
        project: Project,
        modelsProvider: IdeModifiableModelsProvider
    ) {
        if (projectData == null || toImport.isEmpty()) {
            return
        }

        for (node in toImport) {
            val data = node.data
            val module = modelsProvider.findIdeModule(data.module) ?: continue

            val mcpModule = MinecraftFacet.getInstance(module, McpModuleType)
            if (mcpModule != null) {
                // Update the local settings and recompute the SRG map
                mcpModule.updateSettings(data.settings)
                continue
            }

            val children = MinecraftFacet.getChildInstances(module)
            for (child in children) {
                child.getModuleOfType(McpModuleType)?.updateSettings(data.settings)
            }
        }
    }
}
